scripts/checkMesh/checks/collocated_nodes.py

```diff
@@ -0,0 +1,53 @@
+from collections import defaultdict
+from dataclasses import dataclass
+import logging
+from typing import Tuple
+
+import numpy
+
+import vtk  # TODO use new pyvtk style + deal with ImportError
+
+
+@dataclass(frozen=True)
+class Options:
+    tolerance: float
+
+
+@dataclass(frozen=True)
+class Result:
+    nodes_buckets: Tuple[Tuple[int]]
+
+
+def check(vtk_input_file: str, options: Options) -> Result:
+    reader = vtk.vtkXMLUnstructuredGridReader()  # TODO Find a generic way to read the vtk mesh.
+    reader.SetFileName(vtk_input_file)
+    reader.Update()
+    mesh = reader.GetOutput()
+
+    points = mesh.GetPoints()
+
+    locator = vtk.vtkIncrementalOctreePointLocator()
+    locator.SetTolerance(options.tolerance)
+    output = vtk.vtkPoints()
+    locator.InitPointInsertion(output, points.GetBounds())
+
+    # original ids to/from filtered ids.
+    # original_to_filtered = numpy.ones(points.GetNumberOfPoints(), dtype=int) * -1
+    filtered_to_original = numpy.ones(points.GetNumberOfPoints(), dtype=int) * -1
+
+    rejected_points = defaultdict(list)
+    point_id = vtk.reference(0)
```

Comment:
I don't really get what is a `vtk.reference()`, does point_id change within the loop ? If not `filtered_to_original[point_id.get()] = i` always fill the same index of the array. I must have missed something...

Reply:
The doc says that `the index of the point (either new or the duplicate) is returned via point_id.`.
So if the point is inserted, then `point_id` contains the new id of the point (it's been merged, it has a new numbering) and we keep track of the mapping `new_index -> old_index`.

If it's not inserted, then `point_id` contains the node that was already at that location. But `point_id` is the new numbering, in the destination points array. I found most useful for the user to get the old index, so he can look for it in his data.

`point_id` is a `vtk.reference()` because it's written by the `locator.InsertUniquePoint` call.

See https://vtk.org/doc/nightly/html/classvtkIncrementalOctreePointLocator.html

Reply:
Thanks for the explanation!